HW4_Trofimov/das_protein_tools.py

```diff
@@ -0,0 +1,361 @@
+        else:
+            aligned_seq1.append("-")
+            # check original case of amionacid in seq2
+            if seq2[j - 1].isupper():
+                aligned_seq2.append(seq2_upper[j - 1])
+            else:
+                aligned_seq2.append(seq2[j - 1])
+
+            j -= 1
+
+    aligned_seq1 = "".join(reversed(aligned_seq1))
+    aligned_seq2 = "".join(reversed(aligned_seq2))
+
+    return f"{aligned_seq1}, {aligned_seq2}, final score: {dp[m][n]}"
+
+
+# Function to calculate frequency of unique aminoacid in the sequence
+def calculate_aa_freq(sequences: str) -> dict:
+    """
+    Calculates the frequency of each amino acid in a protein sequence or sequences.
+
+    :param sequences: protein sequence or sequences
+    :type sequences: str or list of str
+    :return: dictionary with the frequency of each amino acid
+    :rtype: dict
+    """
+
+    # Creating a dictionary with aminoacid frequencies:
+    amino_acid_frequency = {}
+
+    for amino_acid in sequences:
+        # If the aminoacid has been already in:
+        if amino_acid in amino_acid_frequency:
+            amino_acid_frequency[amino_acid] += 1
+        # If the aminoacid hasn't been already in:
+        else:
+            amino_acid_frequency[amino_acid] = 1
+
+    return amino_acid_frequency
+
+
+# Function to convert one-letter protein sequence to three-letter protein sequence
+def convert_to_3L_code(seq: str) -> str:
+    """
+    This function takes one letter aminoacids sequence and convert's it to three leter coding
+
+    Arguments:
+        seq (str): A sequence of aminoacids
+
+    Output:
+        same sequence but in three-letter coding
+    """
+    seq = seq.upper()
+    if is_protein(seq) is True:
+        sequence = "".join(pd.aa_one_to_three_letter.get(aa) for aa in seq)
+        return sequence[:-1]
+    else:
+        raise ValueError("Sequence is not a protein, input should be protein")
+
+
+# Function to calculate protein mass
+def protein_mass(seq: str) -> float:
+    """
+    This function takes aminoacids sequence and counts it's summary molecular weight using monoisotopic masses
+
+    Arguments:
+        seq (str): A sequence of aminoacids
+
+    Output:
+        returns molecular weight
+    """
+    seq = seq.upper()
+    if is_protein(seq) is True:
+        mass = sum(pd.aa_monoistopic_mass_dict.get(aa) for aa in seq)
+        return mass
+    else:
+        raise ValueError("Sequence is not a protein, input should be protein")
+
+
+# Function to translate Protein to RNA
+def translate_protein_rna(seq: str) -> str:
+    """
+    This function takes  aminoacid sequence and translates in to the RNA.
+    As most of the aminoacids are coded with several different codons,
+    this function will take a random codon of the set for such aminoacids.
+
+    Arguments:
+        seq (str): A sequence of RNA molecule
+
+    Output:
+        returns sequence of aminoacids
+    """
+    seq = seq.upper()
+    if is_protein(seq) is True:
```

Comment:
Как-то много вложенности, как минимум, один уровень убирается обратной проверкой:
```suggestion
    if not is_protein(seq):
        raise ValueError("Sequence is not a protein, input should be a protein")
```